File: cli/src/main/java/nisui/cli/EntryPoint.java
package nisui.cli;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.PrintStream;
import java.io.UnsupportedEncodingException;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import picocli.CommandLine;
import nisui.cli.print_formats.*;
import nisui.core.NisuiFactory;

@CommandLine.Command
public class EntryPoint {
    private NisuiFactory nisuiFactory;

    @CommandLine.Option(names = {"-h", "--help"}, usageHelp = true, description = "Show this help message and exit.")
    private boolean helpRequested;

    @CommandLine.Option(names = {"--format"}, description = "The format to print the results in.")
    String format = "";

    public EntryPoint(NisuiFactory nisuiFactory) {
        this.nisuiFactory = nisuiFactory;
    }

    public void run(InputStream in, PrintStream out, String... args) {
        CommandLine commandLine = new CommandLine(this);

        new ExperimentSubcommand(nisuiFactory).register(commandLine);
        new DataPointSubcommand(nisuiFactory).register(commandLine);
        new ExperimentResultSubcommand(nisuiFactory).register(commandLine);
        new RunSubcommand(nisuiFactory).register(commandLine);

        List<CommandLine> subCommands = commandLine.parse(args);

        if (helpRequested) {
            commandLine.usage(out);
            return;
        }

        boolean ranSomething = false;

        for (int i = 0; i < subCommands.size(); ++i) {
            CommandLine subCommand = subCommands.get(i);
            Object obj = subCommand.<Object>getCommand();
            if (obj instanceof CommandGroup) {
                CommandGroup commandGroup = CommandGroup.class.cast(obj);
                if (
                        commandGroup.helpRequested
                        || i == subCommands.size() - 1
                        || subCommands.get(i + 1).getParent() != subCommand
                   ) {
                    subCommand.usage(out);
                    return;
                }
                switch (format) {
                    case "":
                        commandGroup.printFormatSupplier = PrintTabular::new;
                        break;
                    case "csv":
                        commandGroup.printFormatSupplier = PrintCSV::new;
                        break;
                }
            }
            if (obj instanceof SubCommand) {
                SubCommand subCmd = SubCommand.class.cast(obj);
                if (subCmd.helpRequested) {
                    subCommand.usage(out);
                    return;
                }
                ranSomething = true;
                subCmd.run(in, out);
            }
        }

        if (!ranSomething) {
            commandLine.usage(out);
        }
    }

    public void run(String... args) {
        run(System.in, System.out, args);
    }

    public String runGetOutput(String... args) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        try (PrintStream ps = new PrintStream(baos, true, "utf-8")) {
            run(null, ps, args);
            ps.flush();
            return new String(baos.toByteArray(), StandardCharsets.UTF_8);
        } catch (UnsupportedEncodingException e) {
            throw new Error(e);
        }
    }
}

File: cli/src/main/kotlin/nisui/cli/EntryPoint.kt
package nisui.cli

import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.PrintStream
import java.io.UnsupportedEncodingException
import java.nio.charset.StandardCharsets
import java.util.Arrays
import java.util.List

import picocli.CommandLine

import nisui.cli.print_formats.*
import nisui.core.NisuiFactory

@CommandLine.Command
public class EntryPoint(val nisuiFactory: NisuiFactory) {
    @CommandLine.Option(names = ["-h", "--help"], usageHelp = true, description = ["Show this help message and exit."])
    var helpRequested: Boolean = false

    @CommandLine.Option(names = ["--format"], description = ["The format to print the results in."])
    var format: String = ""

    fun run(in_: InputStream?, out_: PrintStream, vararg args: String) {
        val commandLine = CommandLine(this)

        val mainCommand = StartGuiSubcommand(nisuiFactory)
        mainCommand.register(commandLine)

        ExperimentSubcommand(nisuiFactory).register(commandLine)
        DataPointSubcommand(nisuiFactory).register(commandLine)
        ExperimentResultSubcommand(nisuiFactory).register(commandLine)
        RunSubcommand(nisuiFactory).register(commandLine)
        QueriesSubcommand(nisuiFactory).register(commandLine)

        val subCommands = commandLine.parseArgs(*args).asCommandLineList()

        if (helpRequested) {
            commandLine.usage(out_)
            return
        }

        var ranSomething = false

        for ((i, subCommand) in subCommands.withIndex()) {
            val obj = subCommand.getCommand<Any>()
            if (obj is CommandGroup) {
                val commandGroup: CommandGroup = obj
                if (
                        commandGroup.helpRequested
                        || i == subCommands.size - 1
                        || subCommands.get(i + 1).parent != subCommand
                   ) {
                    subCommand.usage(out_)
                    return
                }
                commandGroup.printFormatCreator = when (format) {
                    "" -> ({ PrintTabular<Any>() })
                    "csv" -> ({ PrintCSV<Any>() })
                    else -> {
                        throw Exception("Unknown format $format")
                    }
                }
            }
            if (obj is SubCommand) {
                val subCmd: SubCommand = obj
                if (subCmd.helpRequested) {
                    subCommand.usage(out_)
                    return
                }
                ranSomething = true
                subCmd.run(in_, out_)
            }
        }

        if (!ranSomething) {
            mainCommand.run(in_, out_)
        }
    }

    fun run(vararg args: String) {
        run(System.`in`, System.`out`, *args)
    }

    fun runGetOutput(vararg args: String): String {
        val baos = ByteArrayOutputStream()
        return PrintStream(baos, true, "utf-8").use { ps ->
            run(null, ps, *args)
            ps.flush()
            String(baos.toByteArray(), StandardCharsets.UTF_8)
        }
    }
}
